jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,738 @@
+     *              tested against the {@link ConnectionMetaData#getRemoteSocketAddress() getRemoteSocketAddress()} of
+     *              {@link Request#getConnectionMetaData()}.
+     */
+    public void exclude(InetAddressPattern... patterns)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        for (InetAddressPattern p : patterns)
+            _predicates.exclude(new InetAddressPatternPredicate(p));
+    }
+
+    /**
+     * Exclude {@link InetAddressPattern} in the conditions to be met
+     * @param patterns String representations of {@link InetAddressPattern}s that are
+     *              tested against the {@link ConnectionMetaData#getRemoteSocketAddress() getRemoteSocketAddress()} of
+     *              {@link Request#getConnectionMetaData()}.
+     */
+    public void excludeInetAddress(String... patterns)
+    {
+        for (String p : patterns)
+            exclude(InetAddressPattern.from(p));
+    }
+
+    /**
+     * {@link IncludeExclude#include(Object) Include} arbitrary {@link Predicate}s in the conditions.
+     * @param predicates {@link Predicate}s that are tested against the {@link Request}.
+     * This method is optimized so that a passed {@link MethodPredicate} or {@link PathSpecPredicate} is
+     * converted to a more efficient {@link #includeMethod(String...)} or {@link #include(PathSpec...)} respectively.
+     */
+    @SafeVarargs
+    public final void include(Predicate<Request>... predicates)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        for (Predicate<Request> p : predicates)
+        {
+            if (p instanceof MethodPredicate methodPredicate)
+                includeMethod(methodPredicate._method);
+            else if (p instanceof PathSpecPredicate pathSpecPredicate)
+                include(pathSpecPredicate._pathSpec);
+            else
+                _predicates.include(p);
+        }
+    }
+
+    /**
+     * {@link IncludeExclude#exclude(Object) Exclude} arbitrary {@link Predicate}s in the conditions.
+     * @param predicates {@link Predicate}s that are tested against the {@link Request}.
+     * This method is optimized so that a passed {@link MethodPredicate} or {@link PathSpecPredicate} is
+     * converted to a more efficient {@link #excludeMethod(String...)} or {@link #exclude(PathSpec...)} respectively.
+     */
+    @SafeVarargs
+    public final void exclude(Predicate<Request>... predicates)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        for (Predicate<Request> p : predicates)
+        {
+            if (p instanceof MethodPredicate methodPredicate)
+                excludeMethod(methodPredicate._method);
+            else if (p instanceof PathSpecPredicate pathSpecPredicate)
+                exclude(pathSpecPredicate._pathSpec);
+            else
+                _predicates.exclude(p);
```

Comment:
Missing branch for `InetAddressPatterPredicate`?

Reply:
There is no optimised `IncludeExcludeSet` for `InetAddressPattern`, so they can just be added to the linear set of all other predicates.

However, inet addresses, probably lend themselves to an optimised set... but only if we have users with many `InetAddressPatternPredicates`

Reply:
"linear set of all other predicates" is basically an anti-pattern for Predicate usage.

Reply:
@joakime Sets of Predicates are not an anti-patterns.  In fact, decomposing complex logic up into collections of predicates is kind of the whole point.   Every AND and OR operation is a list of predicates where the list is size 2.   The `PathMappings` class you developed is essentially a linear list of "predicate" mappings that are applied linearly until one matches, only that it has several optimisation can avoid a linear search if there are certain types of mappings present (e.g. prefix searches can use a `Trie`), but ultimately the class devolves to a list of tests applied in order.

The `IncludeExcludeSet` concept is just another convenient way to standardise and optimise combinations of `Predicate`s to make a users task easier when they are combining multiple concerns.

Ultimately this class could just have a single `setPredicate(Predicate<Request>)` method on it, but then it would not be much of a utility/convenience class.   The whole point is that we provide the structure of include/exclude logic, plus some special purpose predicates, so that we can better understand and optimise the users configuration and have more common code and less complex user predicates.

If users wish to have a single predicate that does everything, then all they need do is `conditionalHandler.include(new MyRequestPredicate())` and they have that behaviour.    Giving them the option to be more structured is not an issue and something well worth doing.